plugins/synaptics-rmi/fu-dump.c

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright (C) 2019 Richard Hughes <richard@hughsie.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include "fu-synaptics-rmi-firmware.h"
+
+static gboolean
+fu_dump_parse (const gchar *filename, GError **error)
+{
+	gchar *data = NULL;
+	gsize len = 0;
+	g_autoptr(GBytes) blob = NULL;
+	g_autoptr(FuFirmware) firmware = fu_synaptics_rmi_firmware_new ();
+	g_autofree gchar *str = NULL;
+	if (!g_file_get_contents (filename, &data, &len, error))
+		return FALSE;
+	blob = g_bytes_new_take (data, len);
+	if (!fu_firmware_parse (firmware, blob, 0, error))
+		return FALSE;
+	str = fu_firmware_to_string (firmware);
+	g_print ("%s", str);
+	return TRUE;
+}
+
+static gboolean
+fu_dump_generate (const gchar *filename, GError **error)
+{
+	const gchar *data;
+	gsize len = 0;
+	g_autoptr(GBytes) blob = NULL;
+	g_autoptr(FuFirmware) firmware = fu_synaptics_rmi_firmware_new ();
+	blob = fu_firmware_write (firmware, error);
+	if (blob == NULL)
+		return FALSE;
+	data = g_bytes_get_data (blob, &len);
+	return g_file_set_contents (filename, data, len, error);
+}
```

Comment:
I'm a little confused the purpose of this tool.  Is it just to make sure that the firmware is properly parsed and that the payload is written in the right format?

I'm just wondering if we need a fwupdtool argument for `parse` that calls plugin vfuncs to tear apart payloads generally.

Reply:
it's what i used for fuzzing the firmware parser as the original code had some *interesting* bugs.

Reply:
Nice!

Reply:
On that note, the fuzzer binary wants to be as small as possible and super quick to start -- so I don't actually thing we want something that's generic and requires hundreds of GObject instantiations.